### PR TITLE
mepo: 0.4.2 -> 1.0

### PR DESCRIPTION
**Pull request by laalsaas**
###### Description of changes

Changelog: https://git.sr.ht/~mil/mepo/refs/1.0
(please note that this is skipping some versions in betwen 0.4.2 and 1.0)

Please review this carefully, as this is my first nixpkgs contribution. Here is a list of things I'm especially not sure about/not 100% happy with but don't know how to do better:
- mepo ships with some helper shellscripts, some of these scripts depend on `jq` being in `$PATH`, one of these scripts depends on `gpspipe` (provided py package gpsd) being in `$PATH`.

  This is solved here by putting both jq and gpsd in the `$PATH` of all scripts containing `jq`. I'm not sure whether this is considered "unclean", as it also puts gpsd into the path of the scripts depending only on jq.

- I haven't been able to add support for the geoclue location source, as this seems to require some black magic dbus fuckery which is beyond my knowledge. If sombody knows how to do this, either go ahead or point me to a ressoucce. currently the corresponding script detects that geoclue is not available and uses the alternatives gpsd and mozilla location services.
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by McSinyx — CHANGES_REQUESTED** [review]
Mepo 1.0 supports multiple menu and geolocation options so should we.  Some new dependencies are also added.  You can start from the following WIP patch of mine:

```diff
From 227f7d61fd0037ea523bf50a016b8c3b371e4334 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Nguy=E1=BB=85n=20Gia=20Phong?= <mcsinyx@disroot.org>
Date: Sat, 8 Oct 2022 23:33:48 +0900
Subject: [PATCH] mepo: 0.4.2 -> 1.0

---
 pkgs/applications/misc/mepo/default.nix | 79 ++++++++++++++++++-------
 pkgs/top-level/all-packages.nix         | 15 ++++-
 2 files changed, 72 insertions(+), 22 deletions(-)

diff --git a/pkgs/applications/misc/mepo/default.nix b/pkgs/applications/misc/mepo/default.nix
index 1afcd3c3f1a7..7f7197ba1688 100644
--- a/pkgs/applications/misc/mepo/default.nix
+++ b/pkgs/applications/misc/mepo/default.nix
@@ -9,26 +9,39 @@
 , SDL2_gfx
 , SDL2_image
 , SDL2_ttf
+, findutils
 , jq
 , ncurses
 , inconsolata-nerdfont
+, menuChoice
+, bemenu
 , dmenu
 , xdotool
-, bemenu
-, withX11 ? false
+, directfb
+, kbd
+, libGL
+, vis
+, gnome
+, withGpsd ? false, gpsd
+, withGeoclue ? true, geoclue2-with-demo-agent
 }:
 
 let
-  menuInputs = if withX11 then [ dmenu xdotool ] else [ bemenu ];
+  menuInputs = {
+    bemenu = [ bemenu ];
+    dmenu = [ dmenu xdotool ];
+    framebuffer = [ directfb kbd libGL vis ];
+    zenity = [ gnome.zenity ];
+  }.${menuChoice};
 in stdenv.mkDerivation rec {
   pname = "mepo";
-  version = "0.4.2";
+  version = "1.0";
 
   src = fetchFromSourcehut {
     owner = "~mil";
     repo = pname;
     rev = version;
-    hash = "sha256-k6YXaqB3EwbDPlTvijZf10q+IYwt4/MiqGXL495KIcY=";
+    hash = "sha256-x1g81J/p7P0w8MOt5Xsmshg/xSl57hBkweZMlsLLWfM=";
   };
 
   nativeBuildInputs = [ pkg-config zig makeWrapper ];
@@ -53,25 +66,51 @@ in stdenv.mkDerivation rec {
   installPhase = ''
     runHook preInstall
 
-    zig build -Drelease-safe=true -Dcpu=baseline --prefix $out install
+    zig build -Drelease-safe=true -Dcpu=baseline --prefix $bin install
+    install -d $man/share/man/man1
+    $bin/bin/mepo -docman > $man/share/man/man1/mepo.1.gz
 
     runHook postInstall
   '';
 
-  postInstall = ''
-    wrapProgram $out/bin/mepo_dl.sh\
-      --suffix PATH : ${lib.makeBinPath [ jq ncurses ]}
-    wrapProgram $out/bin/mepo_ui_helper_menu.sh\
-      --suffix PATH : ${lib.makeBinPath menuInputs}
-    for script in $(grep -l jq out/bin/mepo_ui_menu_*.sh); do
-      wrapProgram $script --suffix PATH : $out/bin:${lib.makeBinPath [ jq ]}
-    done
-    for prog in $out/bin/mepo*; do
-      if [ ! -f $out/bin/.$(basename $prog)-wrapped ]; then
-        wrapProgram $prog --suffix PATH : $out/bin
-      fi
-    done
-  '';
+  postInstall = let
+    inherit (lib) concatStrings makeBinPath optional;
+  in concatStrings ((optional withGeoclue ''
+    substituteInPlace $bin/bin/mepo_ui_menu_user_pin_updater.sh\
+      --replace /usr/libexec ${geoclue2-with-demo-agent}/libexec
+  '') ++ [ ''
+    wrapProgram $bin/bin/mepo_dl.sh\
+      --suffix PATH : ${makeBinPath [ jq ncurses ]}
+    wrapProgram $bin/bin/mepo_ui_central_menu.sh\
+      --suffix PATH : $bin/bin:${makeBinPath [ findutils ]}
+    wrapProgram $bin/bin/mepo_ui_helper_menu.sh\
+      --suffix PATH : ${makeBinPath (menuInputs ++ [ findutils ])}
+    wrapProgram $bin/bin/mepo_ui_menu_dbg_queuedownloadinteractive.sh\
+      --suffix PATH : $bin/bin
+    wrapProgram $bin/bin/mepo_ui_menu_pin_editor.sh\
+      --suffix PATH : $bin/bin
+    wrapProgram $bin/bin/mepo_ui_menu_pref_fontsize.sh\
+      --suffix PATH : $bin/bin
+    wrapProgram $bin/bin/mepo_ui_menu_pref_url.sh\
+      --suffix PATH : $bin/bin:${makeBinPath [ findutils ]}
+    wrapProgram $bin/bin/mepo_ui_menu_pref_zoom.sh\
+      --suffix PATH : $bin/bin
+    wrapProgram $bin/bin/mepo_ui_menu_reposition_nominatim.sh\
+      --suffix PATH : $bin/bin:${makeBinPath [ curl jq ]}
+    wrapProgram $bin/bin/mepo_ui_menu_route_graphhopper.sh\
+      --suffix PATH : $bin/bin:${makeBinPath [ curl jq ]}
+    wrapProgram $bin/bin/mepo_ui_menu_route_overpassrelation.sh\
+      --suffix PATH : $bin/bin:${makeBinPath [ curl findutils jq ]}
+    wrapProgram $bin/bin/mepo_ui_menu_search_nominatim.sh\
+      --suffix PATH : $bin/bin:${makeBinPath [ curl jq ]}
+    wrapProgram $bin/bin/mepo_ui_menu_search_overpass.sh\
+      --suffix PATH : $bin/bin:${makeBinPath [ curl jq ]}
+      wrapProgram $bin/bin/mepo_ui_menu_user_pin_updater.sh\
+      --suffix PATH :\ ${makeBinPath ([ curl jq ] ++ (optional withGpsd gpsd))}
+    wrapProgram $bin/bin/mepo --suffix PATH : $bin/bin
+  '' ]);
+
+  outputs = [ "bin" "out" "man" ];
 
   meta = with lib; {
     description = "Fast, simple, and hackable OSM map viewer";
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index be4c2b3183c9..b22d4b719fac 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -29722,8 +29722,19 @@ with pkgs;
 
   merkaartor = libsForQt5.callPackage ../applications/misc/merkaartor { };
 
-  mepo = callPackage ../applications/misc/mepo { };
-  mepo-x11 = callPackage ../applications/misc/mepo { withX11 = true; };
+  mepo-bemenu = callPackage ../applications/misc/mepo {
+    menuChoice = "bemenu";
+  };
+  mepo-dmenu = callPackage ../applications/misc/mepo {
+    menuChoice = "dmenu";
+  };
+  mepo-framebuffer = callPackage ../applications/misc/mepo {
+    menuChoice = "framebuffer";
+  };
+  mepo-zenity = callPackage ../applications/misc/mepo {
+    menuChoice = "zenity";
+  };
+  mepo = mepo-zenity;
 
   meshcentral = callPackage ../tools/admin/meshcentral { };
 
-- 
2.37.3

```

You might want to note in the long description that the gpsd and geoclue services must be enabled for each of it to work.  I couldn't get `geoclue` to function though, even with the following config:

```nix
services.geoclue2 = {
  enable = true;
  appConfig.mepo = {
    isAllowed = true;
    isSystem = true;
  };
};
```

(TBH I can't get gpsd to work either because my laptop does not have a sensor.)

File: pkgs/applications/misc/bfcal/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+, meson
+, ninja
+, pkg-config
+, wrapQtAppsHook
+, qtbase
+}:
+
+stdenv.mkDerivation rec {
+  pname = "bfcal";
+  version = "1.0";
```
Comment (McSinyx) [inline pkgs/applications/misc/bfcal/default.nix:13]: This seems to be added accidentally.

**Comment by laalsaas** [comment]
Yes, sorry I'm just a little bit confused with my git, give me a couple
of mins.

**Comment by laalsaas** [comment]
Fixed it, once again, sorry for the noise created.

@FRidh @jonringer, I didn't intend to request your review, can i "un-request" it? In any case, feel free to ignore this.

**Comment by laalsaas** [comment]
@McSinyx I got your Patch to work with geoclue. The mepo binary doesn't directly call geoclue, it uses the provided `where-am-i` binary for this. Thus, you need to set the corresponding `appConfig` settings for where-am-i, not mepo. `isSystem` can be false.

**Comment by McSinyx** [comment]
I see, could you please note that in the long description?
It's not trivial to figure out.

**Comment by McSinyx** [comment]
> However, I'm currently still figuring out the handling of some weird
> behaviour when a menuprogram other than menuChoice.
> Maybe the deps of the script shouldn't be appended to $PATH, but instead
> $PATH should be set to the deps?

$PATH can be explicitly set for `mepo_ui_helper_menu.sh` now, however for scripts that depend on one another we have to preserve the original $PATH to allow users to override.

> Also, your patch is missing the `bc` dep for `mepo_ui_helper_pref_pan.sh`

Nice catch!

**Comment by laalsaas** [comment]
Is overriding Mepos default script really a usecase many users have? In
that case, should we maybe provide an additional env variable for
overriding, something like MEPO_PATH or MEPO_SCRIPTS? maybe we could
even ask upstream to support this kind of setup.

**Comment by McSinyx** [comment]
Overriding scripts is an intended use case AFAICT.  I've yet to use mepo
extensively enough to need it, but it's immensely useful in Sxmo.

IIUC the problem here is with menu fallback, which is to be deprecated
in favor of defaulting to zenity: https://todo.sr.ht/~mil/mepo-tickets/46

**Comment by laalsaas** [comment]
I'm currently going through all the scripts and figuring out their
dependencies (coreutils is needed but missing almost everywhere. It's
likely that it's in `$PATH` anyway, but i still consider it cleaner that
way.)

@McSinyx Why have you added the kbd package to the deps of framebuffer menu? For
openvt and chvt the script is explicitly using busybox, or is there
another occasion where it's needed that I'm missing?

Also, the more I'm thinking about it, the less I like the Idea of
including `$PATH`. It simply goes against the nature of stateless
packaging and introduces side effects (which, admittedly, can be desired
for overriding, it's really a two-sided sword and I have yet to come up
with a good solution).

I'm pretty new to nixpkgs, so I might be wrong, but another thing I'm
not 100% happy with, is the handling of the `menuChoice`. The system you
introduced is definetely a cool concept, but I see two minor flaws:
1. I don't like the fact that menuChoice is being passed in as a string,
is there a "safer" way of doing this, such as, speaking in terms of
traditional programming "defining an enum" which contains all possible
options?
2. I'm not sure about that one, but i don't like the fact that
apparently `menuChoice` is mutually exclusive. We don't provide proper
support for use cases where i.e. the user want's to use both dmenu and
bemenu. I'd like it if the user could do something like this in his
`environent.systemPackages`:
`(mepo.override { menuChoice = [ dmenu bemenu ]; })`.
Also, with the current situation, I'm not sure what would happen if
someone installs both mepo-zenity and mepo-dmenu.
On the other hand, with the incoming upstream deprecation of multiple
menu systems, I'm not sure whether it's justified to put too much effort
into this problem.

**Comment by laalsaas** [comment]
I just wanted to share what I've achieved in the meantime. You can have a look at it, comment on it and test it. I didn't test all permutations (menu systems, withgeoclue, withgpsd (can't properly test that either bc of lack of sensor), etc) yet, before i do that I'll need some sleep. When you test it, make sure you don't have an older version of mepo installed in your PATH, else the scripts from the previous version will override those from this package. (That's exactly what I don't like about the approach of including `$PATH` into the scripts.) Cheers!

**Comment by McSinyx** [comment]
> Why have you added the kbd package to the deps of framebuffer menu? For
> openvt and chvt the script is explicitly using busybox, or is there
> another occasion where it's needed that I'm missing?

I wasn't aware of busybox dependency and picked up kbd for them following Alpine.  With busybox added I assume kbd can be removed.

> I don't like the fact that menuChoice is being passed in as a string,
> is there a "safer" way of doing this, such as, speaking in terms of
> traditional programming "defining an enum" which contains all possible
> options?

I don't think there exist an enumeration type, though I am not nix expert.  The map however does limit all possible choices, and nix even provide suggestions upon typos.

> i don't like the fact that apparently `menuChoice` is mutually exclusive. We don't provide proper support for use cases where i.e. the user want's to use both dmenu and
bemenu.

People wanting that, if any, can drop into nix shells, which is not convenient, but I suggest not letting perfect be the enemy of good and solve that when someone with such use case actually asks for it.

File: pkgs/applications/misc/mepo/default.nix [hunk 2]
```diff
@@ -53,31 +69,98 @@ in stdenv.mkDerivation rec {
...
+          isAllowed = true;
+          isSystem = false;
+        };
+      };
+
+      ```
+      Finally, you need to override this package, setting `withGeoclue` or `withGpsd` to true, respectively.
+      (It is also possible to enable and configure both.)
+      To override the package, in your `environment.systemPackages`, list your mepo package of choice as
+      `(mepo.override {withGpsd = true; withGeoclue = true;}`. (only enable what you actually have set up).
```
Comment (McSinyx) [inline pkgs/applications/misc/mepo/default.nix:159]: I suggest making this a bit more concise (I figure you wrote it while sleep-deprived):

```suggestion
      ```
      Finally, you need to override this package, setting `withGeoclue`
      and/or `withGpsd` to true, respectively, e.g. list your mepo package
      of choice as `(mepo.override { withGpsd = true; withGeoclue = true; })`.
```

Note that the overridden declaration also works with nix-shell.

**Review by McSinyx — COMMENTED** [review]
> When you test it, make sure you don't have an older version of mepo installed in your PATH, else the scripts from the previous version will override those from this package. (That's exactly what I don't like about the approach of including $PATH into the scripts.)

I drop into a nix shell (which automatically prepend `${mepo}/bin` to $PATH` and mepo with dmenu or zenity, with and without geoclue, are working as expected.

**Review by McSinyx — APPROVED** [review]

**Comment by laalsaas** [comment]
I had a conversation with SuperSandro and he told me that:

- the postInstall hook is messy and i should just loop over all shell scripts and add the deps to all of them
- for the different backends I should write a nixos module instead of telling users to configure it manually in `longDescription`.

I am going to rewrite this.

**Pull request closed without merging**